katz/src/main/kotlin/katz/typeclasses/Monad.kt

```diff
@@ -0,0 +1,50 @@
+package katz.typeclasses
+
+import java.io.Serializable
+import kotlin.coroutines.experimental.Continuation
+import kotlin.coroutines.experimental.EmptyCoroutineContext
+import kotlin.coroutines.experimental.RestrictsSuspension
+import kotlin.coroutines.experimental.intrinsics.COROUTINE_SUSPENDED
+import kotlin.coroutines.experimental.intrinsics.suspendCoroutineOrReturn
+
+interface Monad<F> : Applicative<F> {
+
+    fun <A, B> flatMap(fa: HK<F, A>, f: (A) -> HK<F, B>): HK<F, B>
+
+    override fun <A, B> ap(fa: HK<F, A>, ff: HK<F, (A) -> B>): HK<F, B> =
+            flatMap(ff, { f -> map(fa, f) })
+}
+
+@RestrictsSuspension
+open class MonadContinuation<F, A>(val M : Monad<F>) : Serializable, Continuation<HK<F, A>> {
+
+    override val context = EmptyCoroutineContext
+
+    override fun resume(value: HK<F, A>) {
+        returnedMonad = value
+    }
+
+    override fun resumeWithException(exception: Throwable) {
+        throw exception
+    }
+
+    internal lateinit var returnedMonad: HK<F, A>
+
+    operator suspend fun <B> HK<F, B>.not(): B = bind { this }
```

Comment:
`await` it's not defined in Kotlin, just as @pakoito it's generally acceptable because C# define it as keyword for their corutines(async/await). but testing a bit around this i cannot properly infix it
```Kotlin
infix suspend fun <B> await(kind: HK<F, B>): B = bind { kind }

val x: Int = await Option(1) //Compiler cannot infer the type and expect varargs, 
```
But we can make an extension function that only be visible inside comprehension

```Kotlin
open class MonadContinuation<F, A>(val M: Monad<F>) : Serializable, Continuation<HK<F, A>> {
...
suspend fun <B>  HK<F, B>.await(): B = bind { this }
...
}
val result = Option.monad().binding {
  val x = Option(1).await()
  val y = Option(1).await()
  yields(x + y)
}

Option(3).await() // not compile await() not defined
```

WDYT guys?